cmd/skopeo/proxy.go

```diff
@@ -0,0 +1,265 @@
+
+	ociManifest, err := manifest.OCI1FromManifest(rawManifest)
+	if err != nil {
+		return err
+	}
+	ociSerialized, err := ociManifest.Serialize()
+	if err != nil {
+		return err
+	}
+
+	w.Header().Set("Content-Length", fmt.Sprintf("%d", len(ociSerialized)))
+	w.WriteHeader(200)
+	_, err = io.Copy(w, bytes.NewReader(ociSerialized))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *proxyHandler) implBlob(w http.ResponseWriter, r *http.Request, digestStr string) error {
+	if err := h.ensureImage(); err != nil {
+		return err
+	}
+
+	_, err := io.Copy(io.Discard, r.Body)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.TODO()
+	d, err := digest.Parse(digestStr)
+	if err != nil {
+		return err
+	}
+	blobr, blobSize, err := (*h.imgsrc).GetBlob(ctx, types.BlobInfo{Digest: d, Size: -1}, h.cache)
+	if err != nil {
+		return err
+	}
+	w.Header().Set("Content-Length", fmt.Sprintf("%d", blobSize))
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.WriteHeader(200)
+	verifier := d.Verifier()
+	tr := io.TeeReader(blobr, verifier)
+	_, err = io.Copy(w, tr)
+	if err != nil {
+		return err
+	}
+	if !verifier.Verified() {
+		return fmt.Errorf("Corrupted blob, expecting %s", d.String())
+	}
+	return nil
+}
+
+// ServeHTTP handles two requests:
+//
+// GET /manifest
+// GET /blobs/<digest>
+// POST /quit
+func (h *proxyHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method == http.MethodPost {
+		if r.URL.Path == "/quit" {
+			w.Header().Set("Content-Length", "0")
+			w.WriteHeader(200)
+			h.shutdown = true
+			return
+		}
+	}
+
+	if r.Method != http.MethodGet {
+		w.Header().Set("Content-Length", "0")
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	if r.URL.Path == "" || !strings.HasPrefix(r.URL.Path, "/") {
+		w.Header().Set("Content-Length", "0")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	var err error
+	if err != nil {
+
+	}
+
+	if r.URL.Path == "/manifest" {
+		err = h.implManifest(w, r)
+	} else if strings.HasPrefix(r.URL.Path, "/blobs/") {
+		blob := filepath.Base(r.URL.Path)
+		err = h.implBlob(w, r, blob)
+	} else {
+		w.Header().Set("Content-Length", "0")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(err.Error()))
+		return
+	}
+}
+
+type sockResponseWriter struct {
+	out     io.Writer
+	headers http.Header
+}
+
+func (rw sockResponseWriter) Header() http.Header {
+	return rw.headers
+}
+
+func (rw sockResponseWriter) Write(buf []byte) (int, error) {
+	return rw.out.Write(buf)
+}
+
+func (rw sockResponseWriter) WriteHeader(statusCode int) {
+	rw.out.Write([]byte(fmt.Sprintf("HTTP/1.1 %d OK\r\n", statusCode)))
+	rw.headers.Write(rw.out)
+	rw.out.Write([]byte("\r\n"))
+}
+
+type proxyOptions struct {
+	global  *globalOptions
+	quiet   bool
+	sockFd  int
+	portNum int
+}
+
+func proxyCmd(global *globalOptions) *cobra.Command {
+	opts := proxyOptions{global: global}
```

Comment:
Should this use `sharedImageFlags`, `imageFlags`, e.g. to allow supplying credentials? (Probably not `retryFlags`.)

(It doesn’t _need_ to be added if you don’t expect to use any of them.)